main.cpp

```diff
@@ -29,25 +29,56 @@ const IR::P4Program *prune_statements(const IR::P4Program *program,
                                       int required_exit_code) {
     cstring stripped_name = remove_extension(options.file);
     stripped_name += "_stripped.p4";
-    for (int j = 0; j < 20; j++) {
-        P4PRUNER::Pruner *pruner = new P4PRUNER::Pruner();
-
+    int same_before_pruning = 0;
+    int max_statements = 100;
+    for (int i = 0; i < 50; i++) {
         auto temp = program;
-        temp = temp->apply(*pruner);
         auto temp_f = new std::ofstream(stripped_name);
+        /* code */
+        // An inspector that collects some statements at random
+        P4PRUNER::Collector *collector =
+            new P4PRUNER::Collector(max_statements);
+        temp->apply(*collector);
+        // Removes all the nodes it recieves from the vector
+        P4PRUNER::Pruner *pruner = new P4PRUNER::Pruner(collector->to_prune);
+        temp = temp->apply(*pruner);
 
         P4::ToP4 *temp_p4 = new P4::ToP4(temp_f, false);
         temp->apply(*temp_p4);
+
         cstring new_command = "python3 ";
         new_command += realpath(options.validator_script, NULL);
         new_command += " -i ";
         new_command += stripped_name;
+        new_command += " 2> /dev/null";
+
         int exit_code = system(new_command.c_str());
 
-        if (exit_code == required_exit_code) {
+        // If we dont see any changes for 7 iterations we probabbly are done
+        if (temp == program) {
+            same_before_pruning++;
+        }
+        if (same_before_pruning >= 7) {
+            break;
+        }
+        /////
```

Comment:
Remove this

Reply:
so sometimes a program is in the smallest form it can be after say 10 iterations, but it would still continue to work for all 50, hence i thought this was a get-around , if a program doesn't change for say 7 iteration, we can probably assume it won't change further hence we don't need to prune this further, this is what i thought

Reply:
Oh I was referring to the ///// comment